### PR TITLE
SPARKC-402: Partition Key Indexes Pushdown Incorrect for PV3

### PR DESCRIPTION
**Pull request by RussellSpitzer**
6f23b30 (Russell Spitzer, 34 minutes ago)
    SPARKC-402: PartitionKey Indexes Pushdown Incorrect for PV3

    In PV3 it is illegal to pushdown an equality predicate based on a secondary
    index for a partition key while also specifying inequalities on the token
    of that partition key. In PV4 it is legal. The pushdowns have now been
    changed to check the PV4 and skip equality pushdowns for partition keys if
    PV3 is in use.

 f82b544 (Russell Spitzer, 8 days ago)
    SPARKC-387: ColumnType Should be ProtocolVersion Aware

    The Connector currently always assumes it is connecting to the newest
    version of C*. This leads to some incorrect decisions when checking against
    ColumnTypes as PV4 Types are not valid in PV3. To fix this all of the
    ColumnType Selection code is now also checks the Protocol Version and
    chooses the Correct Type based Protocol Version instead.

    In addition, the Integeration tests have been version gated to allow for
    accurate testing against previous versions of Cassandra.

### CODE REVIEW

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/RetryUtil.scala [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package com.datastax.spark.connector.util
+
+import scala.annotation.tailrec
+import scala.util.{Failure, Success, Try}
+
+object RetryUtil {
```
Comment (jtgrabowski) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/RetryUtil.scala]: Would [Eventually](http://www.artima.com/docs-scalatest-2.0.M5/org/scalatest/concurrent/Eventually.html) be ok instead?

**Comment by jtgrabowski** [comment]
+1 and minor comment, feel free to merge

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/RetryUtil.scala (hunk 1 above)
Comment (RussellSpitzer) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/RetryUtil.scala]: yep :) I thought I would need something like this in live code, but you are right we don't need this currently.

**Comment by RussellSpitzer** [comment]
@jacek-lewandowski + @jtgrabowski Do you want me to just rebase this and merge it?

**Comment by jtgrabowski** [comment]
I thought that 387 (among other things) introduces tests that fail and this one solves the problem. Let's not ignore 387, let's merge them together.
Are there any showstoppers for 387?

**Comment by RussellSpitzer** [comment]
I don't think so, I"ll do this rebase

**Pull request merged**